File: code/R/ch07/07.08.R
### 7.8.1 Rの場合

library(tidyverse)
library(caret)
my_data <- cars
my_model <- train(form = dist ~ speed, data = my_data, method = "knn")

y_ <- my_model %>% predict(my_data)
RMSE(y_, my_data$speed)
#> [1] 32.24723

f <- function(x) { my_model %>% predict(data.frame(speed = x))}

my_data %>% ggplot(aes(x = speed, y = dist)) +
  geom_point() +         # データ
  stat_function(fun = f) # モデル

File: code/py/ch07/07.08.py
### 7.8.2 Pythonの場合

from sklearn.neighbors import KNeighborsRegressor
my_model = KNeighborsRegressor()

import statsmodels.api as sm
my_data = sm.datasets.get_rdataset('cars', 'datasets').data
X, y = my_data[['speed']], my_data['dist']
my_model.fit(X, y)

y_ = my_model.predict(X)
((y_ - y) ** 2).mean() ** 0.5
#> 13.087184571174962

import pandas as pd
import numpy as np
my_min, my_max = my_data['speed'].min(), my_data['speed'].max()
tmp = pd.DataFrame({'speed':np.linspace(my_min, my_max, num=100)})
tmp['model'] = my_model.predict(tmp)
pd.concat([my_data, tmp]).plot(x='speed', style=['o', '-'])

